File: 121.best-time-to-buy-and-sell-stock.go
/*
 * @lc app=leetcode id=121 lang=golang
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
// Sliding window approach
// T: O(n)
// S: O(1)
func maxProfit(prices []int) int {
    l := 0
	r := 1
	ans := 0
	for r < len(prices) {
		curProfit := prices[r] - prices[l]  // try selling
		if curProfit <= 0 {
			// Buy time and sell time not appropriate, so shift the left boundary  
			// of the window, equivalent to buying at a later time
			l = r
		} else if curProfit > ans {
			// Buy time and sell time yields a positive profit, update the answer
			ans = curProfit
		}
		r++
	}
	return ans
}
// @lc code=end

File: 121.best-time-to-buy-and-sell-stock.cpp
/*
 * @lc app=leetcode id=121 lang=cpp
 *
 * [121] Best Time to Buy and Sell Stock
 */
#include <bits/stdc++.h>

using std::vector;
using std::max;
using std::min;

// @lc code=start
class Solution {
public:
    // Solution 1
    // T: O(n)
    // S: O(1)
    int maxProfit(vector<int>& prices) {
        int ans = 0, buy_price = prices[0];
        for (int i = 1; i < prices.size(); ++i) {
            // Buy stock or not
            buy_price = min(buy_price, prices[i]);

            // Sell stock or not
            ans = max(ans, prices[i] - buy_price);
        }
        return ans;
    }
};
// @lc code=end

